README.md
# Crunker [a fork]

Simple way to merge, concatenate, play, export and download audio files with the Web Audio API.

- No dependencies
- Tiny 2kB gzipped
- Written in Typescript

[View online demos](https://jaggad.github.io/crunker/examples/)

# Installation

```sh
yarn add crunker
```

```sh
npm install crunker
```

# Example

```javascript
let crunker = new Crunker();

crunker
  .fetchAudio("/song.mp3", "/another-song.mp3")
  .then((buffers) => {
    // => [AudioBuffer, AudioBuffer]
    return crunker.mergeAudio(buffers);
  })
  .then((merged) => {
    // => AudioBuffer
    return crunker.export(merged, "audio/mp3");
  })
  .then((output) => {
    // => {blob, element, url}
    crunker.download(output.blob);
    document.body.append(output.element);
    console.log(output.url);
  })
  .catch((error) => {
    // => Error Message
  });

crunker.notSupported(() => {
  // Handle no browser support
});
```

# Condensed Example

```javascript
let crunker = new Crunker();

crunker
  .fetchAudio("/voice.mp3", "/background.mp3")
  .then((buffers) => crunker.mergeAudio(buffers))
  .then((merged) => crunker.export(merged, "audio/mp3"))
  .then((output) => crunker.download(output.blob))
  .catch((error) => {
    throw new Error(error);
  });
```

# Input file Example

```javascript
let crunker = new Crunker();

const onFileInputChange = async (target) => {
  const buffers = await crunker.fetchAudio(
    ...target.files,
    "/voice.mp3",
    "/background.mp3"
  );
};

<input onChange={onFileInputChange(this)} type="file" accept="audio/*" />;
```

## Other Examples

- [Beat Builder Machine](examples/client/beatBuilder.html)

# [Graphic Representation of Methods](https://github.com/jackedgson/crunker/issues/16)

## Merge

![merge](https://user-images.githubusercontent.com/12958674/88806278-968f0680-d186-11ea-9cb5-8ef2606ffcc7.png)

## Concat

![concat](https://user-images.githubusercontent.com/12958674/88806297-9d1d7e00-d186-11ea-8cd2-c64cb0324845.png)

# Methods

For more detailed API documentation, view the Typescript typings.

## new Crunker()

Create a new instance of Crunker.
You may optionally provide an object with a `sampleRate` key, but it will default to the same sample rate as the internal audio context, which is appropriate for your device.

## crunker.fetchAudio(songURL, anotherSongURL)

Fetch one or more audio files.\
**Returns:** an array of audio buffers in the order they were fetched.

## crunker.mergeAudio(arrayOfBuffers);

Merge two or more audio buffers.\
**Returns:** a single `AudioBuffer` object.

## crunker.concatAudio(arrayOfBuffers);

Concatenate two or more audio buffers in the order specified.\
**Returns:** a single `AudioBuffer` object.

## crunker.padAudio(buffer, padStart, seconds);

Pad the audio with silence, at the beginning, the end, or any specified points through the audio.\
**Returns:** a single `AudioBuffer` object.

## crunker.export(buffer, type);

Export an audio buffers with MIME type option.\
**Type:** e.g. `'audio/mp3', 'audio/wav', 'audio/ogg'`.
**IMPORTANT**: the MIME type does **not** change the actual file format. It will always be a `WAVE` file under the hood.\
**Returns:** an object containing the blob object, url, and an audio element object.

## crunker.download(blob, filename);

Automatically download an exported audio blob with optional filename.\
**Filename:** String **not** containing the .mp3, .wav, or .ogg file extension.\
**Returns:** the `HTMLAnchorElement` element used to simulate the automatic download.

## crunker.play(buffer);

Starts playing the exported audio buffer in the background.\
**Returns:** the `HTMLAudioElement`.

## crunker.notSupported(callback);

Execute custom code if Web Audio API is not supported by the users browser.\
**Returns:** The callback function.

# Properties

For more detailed API documentation, view the Typescript typings.

## crunker.context

Access the [AudioContext](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext) used internally by a given Crunker.\
**Returns:** [AudioContext](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext).

# License

MIT


CHANGELOG.md
# 2.2.0 (2023-07-27)


### Bug Fixes

* **ci:** use latest chrome action ([c1a75eb](https://github.com/YumeT023/crunker/commit/c1a75eb0831a7f98cb652c060a873ce4083f5a01))
* corrected minor spelling errors and wrong build script ([ab070e6](https://github.com/YumeT023/crunker/commit/ab070e6a88b913e3dac8e0f22fcacd78be0c8fbd))
* ignore .yarn folder ([3a9bd20](https://github.com/YumeT023/crunker/commit/3a9bd2028fee8dac420f6895b4bebadd1da2cfb9))
* Links to examples in Readme files ([#77](https://github.com/YumeT023/crunker/issues/77)) ([e713c71](https://github.com/YumeT023/crunker/commit/e713c712ea68f58745f98d46eb4c7cc813427406))
* pipeline tests and format prettier; ([7a8982c](https://github.com/YumeT023/crunker/commit/7a8982c2edbd4451e646e581f374fc4a5a677ead))
* update incorrect MIME type default ([46ea851](https://github.com/YumeT023/crunker/commit/46ea8514d85919f660b30389482bf6276bdcc2cc))
* update readme with new npm install info ([ecaf6f8](https://github.com/YumeT023/crunker/commit/ecaf6f8400c492d0d5c5c4401acc692b1897f8b9))
* use latest always for examples ([f607f84](https://github.com/YumeT023/crunker/commit/f607f841cb40e4f2950b04c4cbe7561ba748f01c))


### Features

* add export(...) ([9d2337d](https://github.com/YumeT023/crunker/commit/9d2337db5dc4fb8a2038235d448cf1ae281086be))
* add warning when fetching invalid audio ([f15eae6](https://github.com/YumeT023/crunker/commit/f15eae667cbee06d4a073c841caf2385e087f0e9))
* auto-detect device sample rate if not provided ([#70](https://github.com/YumeT023/crunker/issues/70)) ([f3a27c8](https://github.com/YumeT023/crunker/commit/f3a27c85f73af747cf5f8bb1b17e7f4194f86bc5))
* build ESM module and use in examples ([#87](https://github.com/YumeT023/crunker/issues/87)) ([d41c2ad](https://github.com/YumeT023/crunker/commit/d41c2ad3ad5af43d978df8f94c3da4013e5efd83))
* expose audio context ([71c1f9b](https://github.com/YumeT023/crunker/commit/71c1f9b119cf74263413f21afb0aafd485232e17))
* support Blob in fetchAudio ([#61](https://github.com/YumeT023/crunker/issues/61)) ([03fdfa7](https://github.com/YumeT023/crunker/commit/03fdfa7cb710e41878200b3fee44f49899a4b685))
* upgrade dependencies and fix tests ([83facc5](https://github.com/YumeT023/crunker/commit/83facc5775d960d7a8a280c9099a3078a1224ef4))





test/test.js
import Crunker from "crunker";

describe("Crunker", () => {
  const url = "https://unpkg.com/crunker@1.3.0/examples/server/2.mp3";
  /** @type {Crunker} */
  let audio;
  /** @type {AudioBuffer[]} */
  let buffers;

  beforeEach(async () => {
    audio = new Crunker();
    buffers = await audio.fetchAudio(url, url);
  });

  afterEach(() => {
    audio.close();
  });

  it("creates a context", () => {
    expect(audio._context).to.not.equal(null);
  });

  it("returns internal context", () => {
    expect(audio.context).to.be.instanceOf(window.AudioContext);
  });

  it("should have default sampleRate of 48000", () => {
    expect(audio.context.sampleRate).to.be.eq(48000);
  });

  it("fetches a single audio file", () => {
    expect(buffers[0]).to.have.property("sampleRate", 48000);
  });

  it("fetches multiple audio files", () => {
    buffers.map((buffer) => {
      expect(buffer).to.have.property("sampleRate", 48000);
    });
  });

  it("returns a single buffer when merging", () => {
    expect(audio.mergeAudio(buffers)).to.have.property("sampleRate", 48000);
  });

  it("returns a single buffer when concatenating", () => {
    expect(audio.concatAudio(buffers)).to.have.property("sampleRate", 48000);
  });

  it("uses correct length when concatenating", () => {
    expect(audio.concatAudio(buffers).duration.toFixed(2)).to.equal("16.30");
  });

  it("interleaves two channels", () => {
    const audioInput = buffers[0];
    const interleaved = audio._interleave(audioInput);
    const left = audioInput.getChannelData(0);
    const right = audioInput.getChannelData(1);

    expect(interleaved.length).to.equal(left.length + right.length);
    expect([interleaved[0], interleaved[1]]).to.have.same.members([
      left[0],
      right[0],
    ]);
  });
});


webpack.config.js
const webpack = require("webpack");
const TerserPlugin = require("terser-webpack-plugin");

module.exports = {
  mode: "production",
  context: __dirname + "/src",
  entry: "./crunker.ts",
  output: {
    path: __dirname + "/dist",
    filename: "crunker.js",
    library: "Crunker",
    libraryTarget: "umd",
    umdNamedDefine: true,
  },
  devtool: "source-map",
  module: {
    rules: [
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        loader: "babel-loader",
        options: {
          presets: [
            [
              "@babel/preset-env",
              {
                exclude: [
                  "@babel/plugin-transform-regenerator",
                  "@babel/plugin-transform-async-to-generator",
                ],
              },
            ],
            "@babel/preset-typescript",
          ],
        },
      },
    ],
  },
  optimization: {
    minimize: true,
    minimizer: [new TerserPlugin()],
  },
  plugins: [
    new webpack.DefinePlugin({
      "process.env.NODE_ENV": JSON.stringify(process.env.NODE_ENV),
    }),
  ],
};


webpack.test.config.js
const path = require("path");
const webpack = require("webpack");

const testDir = `${__dirname}/test/`;

module.exports = {
  context: __dirname,
  entry: `${testDir}/test.js`,
  devtool: "inline-source-map",
  output: {
    path: testDir,
    filename: "test.bundle.js",
    library: "Crunker",
    libraryTarget: "umd",
    umdNamedDefine: true,
  },
  resolve: {
    alias: {
      crunker: path.resolve(__dirname, "src/crunker.ts"),
    },
  },
  module: {
    rules: [
      {
        test: /\.(j|t)s$/,
        exclude: /node_modules/,
        loader: "babel-loader",
        options: {
          presets: [
            [
              "@babel/preset-env",
              {
                exclude: [
                  "@babel/plugin-transform-regenerator",
                  "@babel/plugin-transform-async-to-generator",
                ],
              },
            ],
            "@babel/preset-typescript",
          ],
        },
      },
    ],
  },
  plugins: [
    new webpack.DefinePlugin({
      "process.env.NODE_ENV": JSON.stringify(process.env.NODE_ENV),
    }),
  ],
};


src/crunker.ts
export interface CrunkerConstructorOptions {
  /**
   * Sample rate for Crunker's internal audio context.
   *
   * @default 44100
   */
  sampleRate?: number;
}

export type CrunkerInputType = string | File | Blob;

/**
 * Crunker is the simple way to merge, concatenate, play, export and download audio files using the Web Audio API.
 */
export default class Crunker {
  private readonly _sampleRate: number;
  private readonly _context: AudioContext;

  /**
   * Creates a new instance of Crunker with the provided options.
   *
   * If `sampleRate` is not defined, it will auto-select an appropriate sample rate
   * for the device being used.
   */
  constructor(options: CrunkerConstructorOptions = {}) {
    this._context = this._createContext();

    options.sampleRate ||= this._context.sampleRate;

    this._sampleRate = options.sampleRate;
  }

  /**
   * Creates Crunker's internal AudioContext.
   *
   * @internal
   */
  private _createContext(): AudioContext {
    window.AudioContext =
      window.AudioContext ||
      (window as any).webkitAudioContext ||
      (window as any).mozAudioContext;
    return new AudioContext();
  }

  /**
   *
   * The internal AudioContext used by Crunker.
   */
  get context(): AudioContext {
    return this._context;
  }

  /**
   * Asynchronously fetches multiple audio files and returns an array of AudioBuffers.
   */
  async fetchAudio(...sources: CrunkerInputType[]): Promise<AudioBuffer[]> {
    return await Promise.all(
      sources.map(async (source) => {
        let buffer: ArrayBuffer;
        if (source instanceof File || source instanceof Blob) {
          buffer = await source.arrayBuffer();
        } else {
          buffer = await fetch(source).then((response) => {
            if (
              response.headers.has("Content-Type") &&
              !response.headers.get("Content-Type")!.includes("audio/")
            ) {
              console.warn(
                `Crunker: Attempted to fetch an audio file, but its MIME type is \`${
                  response.headers.get("Content-Type")!.split(";")[0]
                }\`. We'll try and continue anyway. (file: "${source}")`
              );
            }
            return response.arrayBuffer();
          });
        }
        return await this._context.decodeAudioData(buffer);
      })
    );
  }

  /**
   * Merges (layers) multiple AudioBuffers into a single AudioBuffer.
   *
   * **Visual representation:**
   *
   * ![](https://user-images.githubusercontent.com/12958674/88806278-968f0680-d186-11ea-9cb5-8ef2606ffcc7.png)
   */
  mergeAudio(buffers: AudioBuffer[]): AudioBuffer {
    const output = this._context.createBuffer(
      this._maxNumberOfChannels(buffers),
      this._sampleRate * this._maxDuration(buffers),
      this._sampleRate
    );

    buffers.forEach((buffer) => {
      for (
        let channelNumber = 0;
        channelNumber < buffer.numberOfChannels;
        channelNumber++
      ) {
        const outputData = output.getChannelData(channelNumber);
        const bufferData = buffer.getChannelData(channelNumber);

        for (
          let i = buffer.getChannelData(channelNumber).length - 1;
          i >= 0;
          i--
        ) {
          outputData[i] += bufferData[i];
        }

        output.getChannelData(channelNumber).set(outputData);
      }
    });

    return output;
  }

  /**
   * Concatenates multiple AudioBuffers into a single AudioBuffer.
   *
   * **Visual representation:**
   *
   * ![](https://user-images.githubusercontent.com/12958674/88806297-9d1d7e00-d186-11ea-8cd2-c64cb0324845.png)
   */
  concatAudio(buffers: AudioBuffer[]): AudioBuffer {
    const output = this._context.createBuffer(
      this._maxNumberOfChannels(buffers),
      this._totalLength(buffers),
      this._sampleRate
    );
    let offset = 0;

    buffers.forEach((buffer) => {
      for (
        let channelNumber = 0;
        channelNumber < buffer.numberOfChannels;
        channelNumber++
      ) {
        output
          .getChannelData(channelNumber)
          .set(buffer.getChannelData(channelNumber), offset);
      }

      offset += buffer.length;
    });

    return output;
  }

  /**
   * Pads a specified AudioBuffer with silence from a specified start time,
   * for a specified length of time.
   *
   * Accepts float values as well as whole integers.
   *
   * @param buffer AudioBuffer to pad
   * @param padStart Time to start padding (in seconds)
   * @param seconds Duration to pad for (in seconds)
   */
  padAudio(
    buffer: AudioBuffer,
    padStart: number = 0,
    seconds: number = 0
  ): AudioBuffer {
    if (seconds === 0) return buffer;

    if (padStart < 0)
      throw new Error(
        'Crunker: Parameter "padStart" in padAudio must be positive'
      );
    if (seconds < 0)
      throw new Error(
        'Crunker: Parameter "seconds" in padAudio must be positive'
      );

    const updatedBuffer = this._context.createBuffer(
      buffer.numberOfChannels,
      Math.ceil(buffer.length + seconds * buffer.sampleRate),
      buffer.sampleRate
    );

    for (
      let channelNumber = 0;
      channelNumber < buffer.numberOfChannels;
      channelNumber++
    ) {
      const channelData = buffer.getChannelData(channelNumber);
      updatedBuffer
        .getChannelData(channelNumber)
        .set(
          channelData.subarray(0, Math.ceil(padStart * buffer.sampleRate) + 1),
          0
        );

      updatedBuffer
        .getChannelData(channelNumber)
        .set(
          channelData.subarray(
            Math.ceil(padStart * buffer.sampleRate) + 2,
            updatedBuffer.length + 1
          ),
          Math.ceil((padStart + seconds) * buffer.sampleRate)
        );
    }

    return updatedBuffer;
  }

  /**
   * Executes a callback if the browser does not support the Web Audio API.
   *
   * Returns the result of the callback, or `undefined` if the Web Audio API is supported.
   *
   * @param callback callback to run if the browser does not support the Web Audio API
   */
  notSupported<T>(callback: () => T): T | undefined {
    return this._isSupported() ? undefined : callback();
  }

  /**
   * Closes Crunker's internal AudioContext.
   */
  close(): this {
    this._context.close();
    return this;
  }

  /**
   * Returns the largest duration of the longest AudioBuffer.
   *
   * @internal
   */
  private _maxDuration(buffers: AudioBuffer[]): number {
    return Math.max(...buffers.map((buffer) => buffer.duration));
  }

  /**
   * Returns the largest number of channels in an array of AudioBuffers.
   *
   * @internal
   */
  private _maxNumberOfChannels(buffers: AudioBuffer[]): number {
    return Math.max(...buffers.map((buffer) => buffer.numberOfChannels));
  }

  /**
   * Returns the sum of the lengths of an array of AudioBuffers.
   *
   * @internal
   */
  private _totalLength(buffers: AudioBuffer[]): number {
    return buffers.map((buffer) => buffer.length).reduce((a, b) => a + b, 0);
  }

  /**
   * Returns whether the browser supports the Web Audio API.
   *
   * @internal
   */
  private _isSupported(): boolean {
    return (
      "AudioContext" in window ||
      "webkitAudioContext" in window ||
      "mozAudioContext" in window
    );
  }

  /**
   * Writes the WAV headers for the specified Float32Array.
   *
   * Returns a DataView containing the WAV headers and file content.
   *
   * @internal
   */
  private _writeHeaders(
    buffer: Float32Array,
    numOfChannels: number,
    sampleRate: number
  ): DataView {
    const bitDepth = 16;
    const bytesPerSample = bitDepth / 8;
    const sampleSize = numOfChannels * bytesPerSample;

    const fileHeaderSize = 8;
    const chunkHeaderSize = 36;
    const chunkDataSize = buffer.length * bytesPerSample;
    const chunkTotalSize = chunkHeaderSize + chunkDataSize;

    const arrayBuffer = new ArrayBuffer(fileHeaderSize + chunkTotalSize);
    const view = new DataView(arrayBuffer);

    this._writeString(view, 0, "RIFF");
    view.setUint32(4, chunkTotalSize, true);
    this._writeString(view, 8, "WAVE");
    this._writeString(view, 12, "fmt ");
    view.setUint32(16, 16, true);
    view.setUint16(20, 1, true);
    view.setUint16(22, numOfChannels, true);
    view.setUint32(24, sampleRate, true);
    view.setUint32(28, sampleRate * sampleSize, true);
    view.setUint16(32, sampleSize, true);
    view.setUint16(34, bitDepth, true);
    this._writeString(view, 36, "data");
    view.setUint32(40, chunkDataSize, true);

    return this._floatTo16BitPCM(
      view,
      buffer,
      fileHeaderSize + chunkHeaderSize
    );
  }

  /**
   * Converts a Float32Array to 16-bit PCM.
   *
   * @internal
   */
  private _floatTo16BitPCM(
    dataview: DataView,
    buffer: Float32Array,
    offset: number
  ): DataView {
    for (let i = 0; i < buffer.length; i++, offset += 2) {
      const tmp = Math.max(-1, Math.min(1, buffer[i]));
      dataview.setInt16(offset, tmp < 0 ? tmp * 0x8000 : tmp * 0x7fff, true);
    }

    return dataview;
  }

  /**
   * Writes a string to a DataView at the specified offset.
   *
   * @internal
   */
  private _writeString(
    dataview: DataView,
    offset: number,
    header: string
  ): void {
    for (let i = 0; i < header.length; i++) {
      dataview.setUint8(offset + i, header.charCodeAt(i));
    }
  }

  /**
   * Exports the specified AudioBuffer to a Blob
   *
   * Note that changing the MIME type does not change the actual file format. The
   * file format will **always** be a WAVE file due to how audio is stored in the
   * browser.
   *
   * @param buffer Buffer to export
   * @param type MIME type (default: `audio/wav`)
   */
  export(buffer: AudioBuffer, type: string = "audio/wav") {
    const recorded = this._interleave(buffer);
    const dataview = this._writeHeaders(
      recorded,
      buffer.numberOfChannels,
      buffer.sampleRate
    );
    return new Blob([dataview], { type });
  }

  /**
   * Converts an AudioBuffer to a Float32Array.
   *
   * @internal
   */
  private _interleave(input: AudioBuffer): Float32Array {
    const channels = Array.from(
      { length: input.numberOfChannels },
      (_, i) => i
    );
    const length = channels.reduce(
      (prev, channelIdx) => prev + input.getChannelData(channelIdx).length,
      0
    );
    const result = new Float32Array(length);

    let index = 0;
    let inputIndex = 0;

    // for 2 channels its like: [L[0], R[0], L[1], R[1], ... , L[n], R[n]]
    while (index < length) {
      channels.forEach((channelIdx) => {
        result[index++] = input.getChannelData(channelIdx)[inputIndex];
      });

      inputIndex++;
    }

    return result;
  }
}
